Add export to CSV functionality in Writer Widget

bruno-pannunzio opened the pull request:
This PR proposes the addition of the functionality to export the phasor table data (with 'G' and 'S') to CSV format.

This functionality is implemented in the Writer Widget, along with the option to export as OME-TIFF.

This was suggested in #55.

zoccoler commented:
Hi Bruno @bruno-pannunzio ,

This is working! Just a couple things:
- if the user changes from ome-tif to csv, maybe the default file name could be updated, because it is always called '... image' (or we leave an empty field if easier)
- to be reproducible, it would be good if the table contained the pixel coordinates in each axis. The current way, the mapping is indirect, going through the 'labels' column, which is like a pixel unique id, but there is no indication where this pixel is in the image unless one looks in the Labels Layer data. I am adding a suggestion for that, but it needs to be expanded to more than 2 dimensions later (tests will likely fail for now due to table shape). Let me know what you think of this, it could be placed in a future release if it now leads to too many changes.
- did you consider at some point using a widget like [FileEdit](https://pyapp-kit.github.io/magicgui/api/widgets/FileEdit/#magicgui.widgets.FileEdit) or [QFileDialog](https://doc.qt.io/qt-6/qfiledialog.html) instead of [QModelIndex](https://doc.qt.io/qt-6/qmodelindex.html) to choose the file path? The way it is right now works fine but, in my case at least, a very large file tree of files and folders gets opened when I use the export widget, which usually collapses the plotter. A FileDialog like widget would put this in another floating window and this would not happen.

bruno-pannunzio commented:
Hi @zoccoler, when you have the chance can you try this? I made some important changes to the 'Phasor Transform' (reader) widget and the Export widget to use the `QFileDialog` instead as suggested, which is a much better implementation. I also think this has great potential for improvement in the future, maybe we can replace completely the widgets for reading and exporting by the default reader and writer by opening the dialog, which can be modified for asking for more inputs (like harmonic, channel, etc). We can discuss this for another release, but it can be a good idea to clean the number of widgets.

Regarding the CSV export. I modified a little bit your implementation for writing the pixel coordinates. I delayed the addition of the coordinates till the moment of exporting, that way we avoid adding more columns to the phasors table, which can be a problem for large images memorywise. Since the pixel coordinates are not used until exporting, there is no need to have this before in the phasors table. But if you see any benefit of having it since the creation of the layer, please let me know. Also, I added the option to store the coordinates for `n` dimensions, considering we plan to support 3D (or any D) in the future, this is a step we are avoiding to change in the future. But we still need to test this thoroughly for more than 2D images.

zoccoler commented:
Hi @bruno-pannunzio ,

It looks much better now! I like the idea of potentially having it as a default reader and writer! Let's keep that in mind!

About the pixel sizes, I agree to adding them only when saving the table. I noticed you added the third dimension (potentially being 'z'). We should have also another one for time. The tricky thing is knowing which dimension is which. This third dimension could mean a 2D time-lapse or a 3D single timepoint. Maybe if we start using [xarrays](https://docs.xarray.dev/en/stable/) for example at some point, we can be more precise and add proper dimension names in the column header. I believe in phasorpy you work with xarrays already, am I right? But we change this in a next release due to time.

I tested this and, before we merge, there are a couple things to consider:

- I applied some threshold to it and then saved both a '.csv' file and an 'ome.tif' file. 
  - In the '.csv', it ended up saving a lot of empty rows (because of the NaNs of the thresholding), which I think could be removed (I don't see any usage for that). This would also help opening the file elsewhere probably (for example, I could not fully open it in Excel as there were "too many rows").
  - I re-opened the 'ome.tif' later with the plugin (drag-and-drop) and plotted the phasor again. What happened is that the threshold was already "applied" even though the threshold slider was at 0. And I could not revert that. On the other hand, I had drawn some clusters before and they were not reloaded (maybe we are always overwriting the 'MANUAL SELECTION #1' column somewhere). I think previous selection columns should be reloaded as they were when saved.
- I tested the 'Phasor Transform' and it worked for a '.ptu' file. But, when I tried to open the recently saved '.ome.tif' file, the button never shows up (or it keeps previous set of widgets from other file types).

bruno-pannunzio commented:
Hi @zoccoler, thanks for all the comments!

> I noticed you added the third dimension (potentially being 'z'). We should have also another one for time. The tricky thing is knowing which dimension is which.

Yes, but actually what I did was to save all the dimensions the image has, so it will also save the time dimension if it has any. Naming the columns can be a little complicated. I think it is possible to save in the metadata when reading the image, but it will take some time to consider all formats and possibilities. We can discuss this for next version as you said. For now I think it's best if the user tracks which dimensions the image has, since they will be in order (the columns are `dim_0`, `dim_1`, etc).

> . Maybe if we start using [xarrays](https://docs.xarray.dev/en/stable/) for example at some point, we can be more precise and add proper dimension names in the column header. I believe in phasorpy you work with xarrays already, am I right?

Actually we dropped Xarray for numpy array in most cases, so I don't know if this will be possible.

> * In the '.csv', it ended up saving a lot of empty rows (because of the NaNs of the thresholding), which I think could be removed (I don't see any usage for that). This would also help opening the file elsewhere probably (for example, I could not fully open it in Excel as there were "too many rows").

I updated the writer widget to drop the rows with `NaN` before saving as '.csv'. I don't know if this can be a problem when recontructing the image, but I think it's a good idea for now. It will make lighter the exported file also.

> * I re-opened the 'ome.tif' later with the plugin (drag-and-drop) and plotted the phasor again. What happened is that the threshold was already "applied" even though the threshold slider was at 0. And I could not revert that. On the other hand, I had drawn some clusters before and they were not reloaded (maybe we are always overwriting the 'MANUAL SELECTION [test commit #1](https://github.com/napari-phasors/napari-phasors/pull/1)' column somewhere). I think previous selection columns should be reloaded as they were when saved.

It will be diffucult to save all the setting from Napari to the exported file. The idea of the export to OME-TIF is to save only the `mean`, `real` and  `imag` (the phasors) of the image after processing it (filtering, thresholding). Saving also the setting for thresholding and filtering can be tricky with the OME-TIF format, since we are a little bit constrained with the format. We could save those setting in a dictionary like format in the field 'description' of the image, but it's not so nice. We can discuss how to save the setting from napari, but the idea of the export to OME-TIF is to further process the phasors in other software, I think. We could save the selections as the csv format, but we would have to implement a reader to reconstruct the image and layer from the csv. This can be a good idea for next release.

> * I tested the 'Phasor Transform' and it worked for a '.ptu' file. But, when I tried to open the recently saved '.ome.tif' file, the button never shows up (or it keeps previous set of widgets from other file types).

Thanks for testing this. I had forgotten to add the 'ome.tif' extension to a advanced widget settings. I think it is fixed now, please let me know if you test it again and it keeps failing.

zoccoler commented:
Hi Bruno, thanks for the reply! I agree with what you said. OK, let's talk about the threshold in the next meeting. 

Just a quick question about the calibration files, since I am acquiring some. I tried to make the phasor plot of the calibration file just to see how it would look like. I was expecting a single cluster somehow on top of the universal semi-circle. But I always get it outside (like below). I tested this with the fluorescein-embryo from FLUTE as well (which is the first image below). Is this to be expected? 

![image](https://github.com/user-attachments/assets/a6c45f36-909e-4150-b8e9-15091988389d)

![image](https://github.com/user-attachments/assets/18e27143-e37d-4015-85f9-7fda486f83f1)

zoccoler commented on src/napari_phasors/_reader.py:
```diff
@@ -318,6 +318,7 @@ def make_phasors_labels_layer(
             harmonic_value = harmonics[0]
         else:
             harmonic_value = harmonics if harmonics is not None else 1
+        # Get pixel coordinates from G_image
```
Please remove this comment.

zoccoler commented:
Hi @bruno-pannunzio ,

The export widget looks much cleaner now, thanks! In general, this looks good to me!

Let's go through this PR tomorrow together to merge it, OK? Just to decide on the threshold subject. Also, we may need to solve some conflicts now that I merged #65
